syncer/streamer_controller.go

```diff
@@ -0,0 +1,343 @@
+
+	streamer         streamer.Streamer
+	streamerProducer StreamerProducer
+
+	// meetError means meeting error when get binlog event
+	// if binlogType is local and meetError is true, then need to create remote binlog stream
+	meetError bool
+
+	fromDB *UpStreamConn
+
+	// changed is true when change from local to remote
+	changed bool
+
+	uuidSuffix string
+}
+
+// NewStreamerController creates a new streamer controller
+func NewStreamerController(tctx tcontext.Context, syncCfg replication.BinlogSyncerConfig, fromDB *UpStreamConn, binlogType BinlogType, localBinlogDir string, timezone *time.Location, beginPos mysql.Position) (*StreamerController, error) {
+	streamerController := &StreamerController{
+		binlogType:     binlogType,
+		syncCfg:        syncCfg,
+		localBinlogDir: localBinlogDir,
+		timezone:       timezone,
+		fromDB:         fromDB,
+	}
+
+	err := streamerController.ResetReplicationSyncer(tctx, beginPos)
+	if err != nil {
+		streamerController.Close(tctx)
+		return nil, err
+	}
+
+	return streamerController, nil
+}
+
+// ResetReplicationSyncer reset the replication
+func (c *StreamerController) ResetReplicationSyncer(tctx tcontext.Context, pos mysql.Position) (err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	var (
+		useRemote            = false
+		uuidSameWithUpstream = true
+	)
+
+	// close old streamerProducer
+	if c.streamerProducer != nil {
+		switch t := c.streamerProducer.(type) {
+		case *remoteBinlogReader:
+			c.closeBinlogSyncer(tctx, t.reader)
+		case *localBinlogReader:
+			uuidSameWithUpstream, err = c.checkUUIDSameWithUpstream(pos, t.reader.GetUUIDs())
+			if err != nil {
+				return err
+			}
+			// TODO: close old local reader before creating a new one
+		default:
+			// some other producers such as mockStreamerProducer, should not re-create
+			c.streamer, err = c.streamerProducer.generateStreamer(pos)
+			return err
+		}
+	}
+
+	// binlog type is local: means relay already use the server id, so need change to a new server id
+	// binlog type is remote: dm-worker has more than one sub task, so need generate random server id
+	randomServerID, err := utils.GetRandomServerID(tctx.Context(), c.fromDB.BaseDB.DB)
+	if err != nil {
+		// should never happened unless the master has too many slave
+		return terror.Annotate(err, "fail to get random server id for streamer controller")
+	}
+	c.syncCfg.ServerID = randomServerID
+
+	// re-create new streamerProducer
+	if c.binlogType == RemoteBinlog {
+		useRemote = true
+	} else {
+		if c.meetError {
+			// meetError is true means meets error when get binlog event, in this case use remote binlog as default
+			if !uuidSameWithUpstream {
+				// if the binlog position's uuid is different from the upstream, can not switch to remote binlog
+				tctx.L().Info("may switch master in upstream, so can not switch local to remote")
+			} else {
+				useRemote = true
+				c.changed = true
+				tctx.L().Info("meet error when read from local binlog, will switch wo remote binlog")
```

Comment:
"will switch wo remote binlog" -- What is "wo"?

Reply:
typo, it `to`, fix in e9e4868